# HG changeset patch
# User Alessandro Castellani <alessandro@thunderbird.net>
# Date 1670857631 -3600
# Node ID 85dcc90b5150f0e291c57d302caa461479a37158
# Parent  20e747200e2fa5e7393f2fd976187d76078f829b
Bug 1801200 - Open the End of Year appeal page in a browser on startup. r=darktrojan  a=rjl

We're opening this in the browser rather than the tab because the donation link does funky things inside Thunderbird.

Differential Revision: https://phabricator.services.mozilla.com/D162898

diff --git a/mail/app/profile/all-thunderbird.js b/mail/app/profile/all-thunderbird.js
--- a/mail/app/profile/all-thunderbird.js
+++ b/mail/app/profile/all-thunderbird.js
@@ -1325,8 +1325,13 @@ pref("mail.useNewMailTabs", false);
 // Hide it and move the title to the center.
 pref("print.print_headerleft", "");
 pref("print.print_headercenter", "&T");
 pref("print.print_headerright", "");
 
 // Enable Masonry Layout for AddressBook.
 pref("layout.css.grid-template-masonry-value.enabled", true);
 
+// End of year donation appeal.
+pref("app.donation.eoy.version", 1);
+pref("app.donation.eoy.version.viewed", 0);
+pref("app.donation.eoy.url", "https://www.betterbird.eu/end-of-year/");
+
diff --git a/mail/base/content/msgMail3PaneWindow.js b/mail/base/content/msgMail3PaneWindow.js
--- a/mail/base/content/msgMail3PaneWindow.js
+++ b/mail/base/content/msgMail3PaneWindow.js
@@ -738,16 +738,22 @@ var gMailInit = {
       gSpacesToolbar.onLoad();
     }
 
     // All core modal dialogs are done, the user can now interact with the
     // 3-pane window. We need to notify this even if the user didn't setup any
     // mail account in order to trigger all the other areas of the application.
     Services.obs.notifyObservers(window, "mail-startup-done");
 
+    // Show the end of year donation appeal page.
+    if (this.shouldShowEOYDonationAppeal()) {
+      // Add a timeout to prevent opening the browser immediately at startup.
+      setTimeout(this.showEOYDonationAppeal, 2000);
+    }
+
     // Idle dispatch the telemetry reports.
     Services.tm.idleDispatchToMainThread(() => {
       reportAccountTypes();
       reportAddressBookTypes();
       reportAccountSizes();
       reportBooleanPreferences();
     });
   },
@@ -789,16 +795,54 @@ var gMailInit = {
     MailServices.mailSession.RemoveFolderListener(folderListener);
 
     // FIX ME - later we will be able to use onload from the overlay
     OnUnloadMsgHeaderPane();
 
     UnloadPanes();
     OnMailWindowUnload();
   },
+
+  /**
+   * Check if we can trigger the opening of the donation appeal page.
+   *
+   * @returns {boolean} - True if the donation appeal page should be opened.
+   */
+  shouldShowEOYDonationAppeal() {
+    let currentEOY = Services.prefs.getIntPref("app.donation.eoy.version", 1);
+    let viewedEOY = Services.prefs.getIntPref(
+      "app.donation.eoy.version.viewed",
+      0
+    );
+
+    // True if the user never saw the donation appeal, this is not a new
+    // profile (since users are already prompted to donate after downloading),
+    // and we're not running tests.
+    return (
+      viewedEOY < currentEOY &&
+      !specialTabs.shouldShowPolicyNotification() &&
+      !Cu.isInAutomation
+    );
+  },
+
+  /**
+   * Open the end of year appeal in a new web browser page. We don't open this
+   * in a tab due to the complexity of the donation site, and we don't want to
+   * handle that inside Thunderbird.
+   */
+  showEOYDonationAppeal() {
+    let url = Services.prefs.getStringPref("app.donation.eoy.url");
+    let messenger =
+      window.messenger ||
+      Cc["@mozilla.org/messenger;1"].createInstance(Ci.nsIMessenger);
+    messenger.launchExternalURL(url);
+
+    let currentEOY = Services.prefs.getIntPref("app.donation.eoy.version", 1);
+    Services.prefs.setIntPref("app.donation.eoy.version.viewed", currentEOY);
+  },
 };
 
 /**
  * Called at startup to verify if we have ny existing account, even if invalid,
  * and if not, it will trigger the Account Hub in a tab.
  *
  * @returns {boolean} - True if we have at least one existing account.
  */
